### DISCUSSION TITLE
I'm using App route, and then I want to implement a flat route file for the API folder

### DISCUSSION THREAD

**Discussion by epver** [vercel/next.js#57754] [App Router]
### Goals

1. api folders can be managed flat

### Non-Goals

1.
2.
3.

### Background

I think the API folder should be done flat processing, if the folder depth is too much, in fact, it is extremely inconvenient to manage

### Proposal

1. folder 
```
/app/
  /api/
    [[...paths]]
      route.js
    login.js
    logout.js
    users.$id.js
    users.js
    projects.$id.js
    projects.$id.me.js
    projects.js
```

I think this makes it easier to manage the content of the API，
Then I have the following implementation and find that import does not asynchronously load the contents of the api folder

### route.js
```javascript
import { readdirSync } from 'node:fs';
import { join } from 'node:path';
import { routeMatcher } from 'route-matcher';

const folder = new URL(join('..'), import.meta.url);
const apiFiles = readdirSync(folder);
const apiRoutes = [];

for (const filename of apiFiles) {
  if (filename.startsWith('_') || filename.startsWith('.')) continue;
  if (!filename.endsWith('.js')) continue;
  const [id] = filename.split('.js');
  const file = `../${filename}`;
  const path = `/api/${id.replace(/\.{1}/gi, '/').replace(/\$/gi, ':')}`;
  const matcher = routeMatcher(path);
  // const resolver = await import(join(import.meta.url, `../${file}`));
  apiRoutes.push({ id, file, path, matcher });
}

export const GET = async (request) => {
  const { pathname } = request.nextUrl;
  console.log(pathname, paths);
  let foundRoute;
  for (const apiRoute of apiRoutes) {
    const params = apiRoute.matcher.parse(pathname);
    if (params) {
      foundRoute = { ...apiRoute, params };
      break;
    }
  }

  if (!foundRoute) {
    return Response.json({ status: 404, message: 'NotFound -- file not create' }, { status: 404 });
  }
  try {
    console.log(foundRoute.file);
    const resolver = await eval(`import('${foundRoute.file}')`);
    if (typeof resolver.GET === 'function') {
      return resolver.GET(request, foundRoute.params ?? {});
    }
  } catch (error) {
    return Response.json({ status: 404, message: 'NotFound -- file content empty' }, { status: 404 });
  }
  // return Response.json({});
};

```

### users.$id.js
```javascript
export async function GET(request, params) {
  return Response.json({ params });
}
```

I know that another solution is to load all the api files into the content first, by directly writing the import way.

**Comment by icyJoseph** [comment]
Someone already created this:

- https://github.com/finom/next-wednesday
- https://github.com/vercel/next.js/discussions/57604 introduction to the community

A controller that can take care of routes by using decorators, and an single catch all API route handler.

I guess the main issue might be having tons of controllers in the same file, which might lead to poor performance of a request/response combo. That'd also be a problem for flat routing approach, within an edge/serverless world of course, since the lambda would have to boot the entire thing on every request...

**Reply by epver** [reply]
I know how to do this, but what I want is to dynamically load the API, and only call the API when it's requested, right

So the above solution is not fit me,  and then I checked the documentation and found that I could not write the plug-in

**Comment by epver** [comment]
https://github.com/ifyio/next-flat-routes This project it's what I want, but the implementation is a little hacker
